moto/glue/utils.py

```diff
@@ -0,0 +1,378 @@
+import abc
+import operator
+import re
+import warnings
+from datetime import date, datetime, timedelta
+from typing import Any, Dict, List, Optional, Union
+
+from pyparsing import (
+    CaselessKeyword,
+    Forward,
+    OpAssoc,
+    ParserElement,
+    ParseResults,
+    QuotedString,
+    Suppress,
+    Word,
+    alphanums,
+    delimited_list,
+    exceptions,
+    infix_notation,
+    one_of,
+    pyparsing_common,
+)
+
+from .exceptions import InvalidInputException, InvalidStateException
+
+
+def _cast(type_: str, value: Any) -> Union[date, datetime, float, int, str]:
+    # values are always cast from string to target type
+    value = str(value)
+
+    if type_ in ("bigint", "int", "smallint", "tinyint"):
+        try:
+            return int(value)  # no size is enforced
+        except ValueError:
+            raise ValueError(f'"{value}" is not an integer.')
+
+    if type_ == "decimal":
+        try:
+            return float(value)
+        except ValueError:
+            raise ValueError(f"{value} is not a decimal.")
+
+    if type_ in ("char", "string", "varchar"):
+        return value  # no length is enforced
+
+    if type_ == "date":
+        try:
+            return datetime.strptime(value, "%Y-%m-%d").date()
+        except ValueError:
+            # NOTE AWS accepts up until 31st of every month, e.g. '2022-02-31'
+            warnings.warn("Date filtering beyond last of month not supported")
+            raise ValueError(f"{value} is not a date.")
+
+    if type_ == "timestamp":
+        match = re.search(
+            r"^(?P<timestamp>\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2})"
+            r"(?P<nanos>\.\d{1,9})?$",
+            value,
+        )
+        if match is None:
+            raise ValueError(
+                "Timestamp format must be yyyy-mm-dd hh:mm:ss[.fffffffff]"
+                f" {value} is not a timestamp."
+            )
+
+        try:
+            timestamp = datetime.strptime(match.group("timestamp"), "%Y-%m-%d %H:%M:%S")
+        except ValueError:
+            # NOTE AWS accepts up until 31st of every month, e.g. '2022-02-31 00:00:00'
+            warnings.warn("Timestamp filtering beyond last of month not supported")
+            raise ValueError(
+                "Timestamp format must be yyyy-mm-dd hh:mm:ss[.fffffffff]"
+                f" {value} is not a timestamp."
+            )
+
+        nanos = match.group("nanos")
+        if nanos is not None:
+            # strip leading dot, reverse and left pad with zeros to nanoseconds
+            nanos = "".join(reversed(nanos[1:])).zfill(9)
+            for i, nanoseconds in enumerate(nanos):
+                # NOTE precision loss here, as nanoseconds are not supported in datetime
+                microseconds = (int(nanoseconds) * 10**i) / 1000
+                if round(microseconds) == 0 and microseconds > 0:
+                    warnings.warn("Nanoseconds not supported, rounding to microseconds")
```

Comment:
Considering Moto will never deal with nanoseconds, I don't think this warning is necessary. Let's just treat it as an implementation detail. It should be part of the documentation in the docstring of `get_partition`, but we don't have to warn the user everytime.